### DISCUSSION TITLE
Enable to select models in Copilot

### DISCUSSION THREAD

**Discussion by ryoppippi** [olimorris/codecompanion.nvim#385] [Ideas]
Hi!
In recent update, we can choose multiple models on Copilot!
I'd like to use it in codecompanion just like we can choose models on ollama
![SCR-20241101-kixd](https://github.com/user-attachments/assets/a9bc113d-fbc8-4a89-8f73-e5411ca76569)

**Comment by olimorris** [comment]
I'd welcome a PR for this. You can update the `schema.model.choices` in the adapter.

**Comment by olimorris** [comment]
Do you know the full model names?

**Reply by knoc-off** [reply]
i believe this is what your looking for?

  1: gpt-3.5-turbo-0613
  2: gpt-4-0613
  3: gpt-4o-2024-05-13
  4: gpt-4o-2024-08-06
  5: o1-preview-2024-09-12
  6: o1-mini-2024-09-12
  7: gpt-4o-mini-2024-07-18
  8: gpt-4-0125-preview
  9: claude-3.5-sonnet

i got this using the CopilotChatModels, from this plugin, source shown below.

https://github.com/CopilotC-Nvim/CopilotChat.nvim/blob/76bfba5cd8915ff55554e0649809bec0207b6a83/lua/CopilotChat/init.lua#L902-L1001

snippet from init.lua:
```
--- Select a Copilot GPT model.
function M.select_model()
  async.run(function()
    local models = state.copilot:list_models()

    vim.schedule(function()
      vim.ui.select(models, {
        prompt = 'Select a model',
      }, function(choice)
        if choice then
          M.config.model = choice
        end
      end)
    end)
  end)
end
```

hope this helps

**Reply by olimorris** [reply]
Just wanted to add my thanks to @knoc-off. Really appreciated. I've made some progress on the o1 models that hopefully I can share tonight.

**Comment by ryoppippi** [comment]
Actually there is no full model list.
I can use sonnet, but other people can use perplexity.
I'll add models which I can use, but more contributors add them.

**Comment by AntoineBalaine** [comment]
Sorry to revive this closed discussion, but could you explain how selecting models ought to be done in the config?
In a Lazyvim config: trying to setup as follows:
```lua
return {
  {
    "olimorris/codecompanion.nvim",
    dependencies = {
      "nvim-lua/plenary.nvim",
      "nvim-treesitter/nvim-treesitter",
    },
    config = function()
      require("codecompanion").setup({
        strategies = {
          chat = { adapter = "copilot" },
          inline = { adapter = "copilot" },
        },
        opts = {
          log_level = "DEBUG",
        },
      })
    end,
  },
}
```
…I’m blocked figuring out what I’m supposed to do beyond this - I see the models are listed in the copilot adapters, but I’m not quite clear what I must do to pick which model to use.

Other question (lmk if this requires a separate thread): is there any way to retrieve a chat history and its context? I reckon it’s probably possible, but I can’t find how to do it either.

**Reply by olimorris** [reply]
Please see the "Changing an Adapter's Default Model" in the README.

**Reply by AntoineBalaine** [reply]
Thank you.

**Reply by ghost** [reply]
Were you able to solve the problem? I somehow can't and the docs are kind of confusing...

**Reply by olimorris** [reply]
> Were you able to solve the problem? I somehow can't and the docs are kind of confusing...

What's confusing you?

**Reply by olimorris** [reply]
Quick tip, if any docs don't make sense, search GitHub for other people's config. You can learn a lot about how they're setting things up
